src/vidpf.rs

```diff
@@ -0,0 +1,828 @@
+        );
+        let mut rng = shake.into_seed_stream();
+
+        let mut proof = Proof::default();
+        rng.fill_bytes(&mut proof.0);
+
+        Ok(proof)
+    }
+
+    fn node_proof_adjustment(mut proof: Proof) -> Proof {
+        let mut rng = XofTurboShake128::seed_stream(
+            &XofSeed(Default::default()),
+            DomainSepTag::NODE_PROOF_ADJUST,
+            &proof.0,
+        );
+        rng.fill_bytes(&mut proof.0);
+
+        proof
+    }
+}
+
+/// Contains the domain separation tags for invoking different oracles.
+struct DomainSepTag;
+impl DomainSepTag {
+    const PRG: &'static [u8] = b"Prg";
+    const CONVERT: &'static [u8] = b"Convert";
+    const NODE_PROOF: &'static [u8] = b"NodeProof";
+    const NODE_PROOF_ADJUST: &'static [u8] = b"NodeProofAdjust";
+}
+
+/// [`Key`] is the private key of an aggregation server.
+pub struct Key {
+    id: ServerId,
+    value: [u8; 16],
+}
+
+impl Key {
+    /// Generates a key at random.
+    ///
+    /// # Errors
+    /// Triggers an error if the random generator fails.
+    pub(crate) fn gen(id: ServerId) -> Result<Self, VidpfError> {
+        let mut value = [0; 16];
+        getrandom::getrandom(&mut value)?;
+        Ok(Self { id, value })
+    }
+}
+
+/// `ServerId` used to identify the two aggregation servers.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(crate) enum ServerId {
+    /// S0 is the first server.
+    S0,
+    /// S1 is the second server.
+    S1,
+}
+
+impl From<ServerId> for Choice {
+    fn from(value: ServerId) -> Self {
+        match value {
+            ServerId::S0 => Self::from(0),
+            ServerId::S1 => Self::from(1),
+        }
+    }
+}
+
+/// [`CorrectionWord`] adjusts values of shares during the VIDPF evaluation.
+struct CorrectionWord<W: VidpfWeight> {
+    seed: Seed,
+    left_control_bit: Choice,
+    right_control_bit: Choice,
+    weight: W,
+}
+
+/// [`PublicShare`] is the common public information used by aggregation servers.
+pub struct PublicShare<W: VidpfWeight> {
+    cw: Vec<CorrectionWord<W>>,
+    cs: Vec<Proof>,
+}
+
+/// [`EvalState`] contains the values produced during input evaluation at a given level.
+pub struct EvalState {
+    seed: Seed,
+    control_bit: Choice,
+    proof: Proof,
+}
+
+impl EvalState {
+    fn init_from_key(key: &Key) -> Self {
+        Self {
+            seed: key.value,
+            control_bit: Choice::from(key.id),
+            proof: Proof::default(),
+        }
+    }
+}
+
+/// [`ValueShare`] contains a share of the input's weight together with a proof for verification.
+pub struct ValueShare<W: VidpfWeight> {
+    /// Secret share of the input's weight.
+    pub share: W,
+    /// Proof used to verify the share.
+    pub proof: Proof,
+}
+
+/// Proof size in bytes.
+const VIDPF_PROOF_SIZE: usize = 32;
+
+/// [`Proof`] is used to validate user input and shares after evaluation.
+#[derive(Debug, PartialEq, Eq, Default)]
+pub struct Proof([u8; VIDPF_PROOF_SIZE]);
```

Comment:
This should probably be crate-private. Also, consider whether this could be just a type alias for `[u8; VIDPF_PROOF_SIZE]` rather than a wrapper for it.

```suggestion
pub(crate) struct Proof([u8; VIDPF_PROOF_SIZE]);
```